Fetch to upstream

xagent003 opened the pull request.

mithilarun commented:
From a working calico-kubeconfig on a IPv4 host:
```
- name: local
  cluster:
    server: https://[10.149.0.1]:443
```

e0ne committed:
Use k8snetworkplumbingwg repo instead of dougbtv in sources

rthakur-est committed:
propagate context timeout error

When context timeout occurs while updating the ip pool CRD,
error is not returned but allocated IP is returned. This leads
to duplicate IP allocation.

This change fixes the above by returning error.

Signed-off-by: Ravindra Thakur <ravindra.nath.thakur@est.tech>

martinkennelly committed:
Add total Add/Del timeout

During scale testing, we found that if an instance
of whereabouts did not accomplish its task within a
period of time, the kubelet spawned an additional
whereabouts instance and the initial instance
remained active.

For pod creation, the Kubelet will try to
start a pod and wait for ~8 minutes before
attempting to kill it. See kubelets "runPod"
function.

For pod deletion, kubelet will attempt to kill a pod
and if it fails it will attempt after one minute one
more time. Unable to see this in kubelet code but
observed it using K8 1.22.

Signed-off-by: Martin Kennelly <mkennell@redhat.com>

martinkennelly committed:
Release lease lock when LE ends

Enable LE config attribute 'ReleaseOnCancel'
which will cause release of the associated
lock when LE ends.

Signed-off-by: Martin Kennelly <mkennell@redhat.com>

xagent003 committed:
ip-reconciler: Add all non default interfaces to Pod IP list

martinkennelly committed:
Add E2E test cases and execute with GH workflow

Test cases are simple and check that:
* IPs seen in IP pool are what is assigned
to pods.
* No stale IPs in IP pool following test run.

Signed-off-by: Martin Kennelly <mkennell@redhat.com>

antoninbas committed:
Upload whereabouts binary as release asset

The binaries (built for amd64, arm64 and arm) are uploaded using a
Github workflow every time a new Github release is created. This
simplifies consumption of these binaries by other projects, and creates
an "official" build.

Signed-off-by: Antonin Bas <abas@vmware.com>

dhiller committed:
Add label app=whereabouts to ip-reconciler pod template

This will allow excluding the pod that is created in the default
namespace i.e.

kubectl  wait --for=condition=Ready pod --timeout=30s -l app!=whereabouts

Signed-off-by: Daniel Hiller <dhiller@redhat.com>

dougbtv committed:
Merge pull request #149 from antoninbas/upload-whereabout-binary-as-release-asset

Upload whereabouts binary as release asset

dougbtv committed:
Merge pull request #143 from Nordix/ctx_timeout

propagate context timeout error

dougbtv committed:
Merge pull request #145 from martinkennelly/enable_release_cancel

Release lease lock when LE ends

dougbtv committed:
Merge pull request #142 from martinkennelly/fix_multi_instance

Add total Add/Del timeout

dougbtv committed:
Merge pull request #148 from martinkennelly/add_simple_e2e_test

Add E2E test cases and execute with GH workflow

nicklesimba committed:
added sleep-for-race capability for testing

martinkennelly committed:
Fix selection to include only test pods

Signed-off-by: Martin Kennelly <mkennell@redhat.com>

martinkennelly committed:
Throw error when IP pool and pod IP total mismatch

Signed-off-by: Martin Kennelly <mkennell@redhat.com>

dougbtv committed:
Merge pull request #151 from nicklesimba/sleep-for-race

added sleep-for-race option for IPAM

dougbtv committed:
Merge pull request #152 from martinkennelly/fix_test

Fix test to include error when IP pool total doesn't equal assigned WB pod IPs

maiqueb committed:
build: enable e2e testing for podman

As per [0] kind `load docker-image` commands are docker specific.
This commit changes that instruction, using the generic
`load image-archive` instead, which works both for podman and docker.

[0] - https://github.com/kubernetes-sigs/kind/issues/2027

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
e2e tests: make WB play along with containerd

containerd sends CNI CMD check instructions to the plugins, and
whereabouts does not implement that command, returning with an
error.

As per the CNI specification - [0] - the runtime "may assume that
a failed CHECK means the container is permanently in a misconfigured
state." and as a results replies with a CNI del.

By instructing the CRI to not send CNI checks we work around this
issue.

[0] - https://github.com/containernetworking/cni/blob/master/SPEC.md#check-check-containers-networking-is-as-expected

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
e2e tests: manually reconcile the cluster

Before checking for stale IP addresses, trigger the `ip-reconciler`
tool, which will cleanup stale IP addresses.

In order to manually run the reconciler, we need to build it first;
as such, the github action must be updated.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
build, ip reconciler: have configurable logging

The ip-reconciler tool currently feature the logging level
hard-coded to `debug`, which is too verbose.

This commit makes it configurable, defaulting it to error level.

The effective logging level is set to `verbose` in the daemonset
spec.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

dougbtv committed:
Merge pull request #154 from maiqueb/e2e-kind-podman-support

build: enable e2e testing for podman

dougbtv committed:
Merge pull request #155 from maiqueb/configurable-log-level-reconciler

build, ip reconciler: have configurable logging

dougbtv committed:
Merge pull request #123 from e0ne/k8snetworkplumbingwg-group

Use k8snetworkplumbingwg repo instead of dougbtv in sources

maiqueb committed:
reconciler, spec: set priority class

Define the reconciler as a system critical resource.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
reconciler, spec: remove the resource limits

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

dougbtv committed:
Merge pull request #158 from maiqueb/fix-reconciler-spec

Fix reconciler spec

maiqueb committed:
build: install kubebuilder tools for podman

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
ip-reconciler: run without a defined kubeconfig

With this change the reconciler can run in two different ways:
- when ran in a k8s pod, it does not require to be told how to
  connect to the cluster.
- when ran as a binary, it does require to know how to connect to
  the cluster, via the -kubeconfig config option.

The reconciler cron spec is updated to use the correct service
account name, and also is updated to run in the `kube-system`
namespace.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

maiqueb committed:
reconciler: update logging call

All (non error) logs have the same verbosity; it would be interesting
to see the reconciled IPs by default, as such I'm bumping this
particular logging call to `verbose`.
That enables users to `kubectl logs <reconciler-pod-name` and see which
IP addresses got reconciled.

The below calls would be bumped to `verbose`.
```
2021-10-20T15:50:00Z [debug] removed stale overlappingIP allocation [10.10.0.100]
2021-10-20T15:50:00Z [debug] removed stale overlappingIP allocation [10.10.0.111]
2021-10-20T15:50:00Z [debug] removed stale overlappingIP allocation [10.10.0.113]
2021-10-20T15:50:00Z [debug] removed stale overlappingIP allocation [10.10.0.118]
2021-10-20T15:50:00Z [debug] removed stale overlappingIP allocation [10.10.0.13]
2021-10-20T15:50:00Z [debug] removed stale overlappingIP allocation [10.10.0.170]
2021-10-20T15:50:00Z [debug] removed stale overlappingIP allocation [10.10.0.18]
2021-10-20T15:50:00Z [debug] removed stale overlappingIP allocation [10.10.0.3]
2021-10-20T15:50:00Z [debug] removed stale overlappingIP allocation [10.10.0.35]
2021-10-20T15:50:00Z [debug] removed stale overlappingIP allocation [10.10.0.4]
2021-10-20T15:50:01Z [debug] removed stale overlappingIP allocation [10.10.0.83]
```

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

dougbtv committed:
Merge pull request #159 from maiqueb/remove-kubeconfig-from-k8s-client

ip-reconciler: remove kubeconfig from k8s client

dougbtv committed:
Merge pull request #156 from maiqueb/disable-cni-check-cmd

e2e tests: fix currently broken setup on kind clusters

nicklesimba committed:
Configurable sleep_for_race parameter

dougbtv committed:
Merge pull request #161 from nicklesimba/sleep-for-race

Configurable sleep_for_race parameter

maiqueb committed:
ip-reconciler: do not reconcile pods in Pending phase

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

dougbtv committed:
Merge pull request #150 from dhiller/add-app-label-to-cronjob-pod

Add label app=whereabouts to ip-reconciler pod template

dougbtv committed:
Merge pull request #147 from xagent003/private/arjun/ipReconcilerNetName

ip-reconciler: Add all non default interfaces to Pod IP list

dougbtv committed:
Merge pull request #166 from maiqueb/skip-reconciler-pending-pods

ip-reconciler: do not reconcile pods in Pending phase

maiqueb committed:
Safer reconciler spec (#167)

* reconciler, spec: forbid concurrent executions

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

* reconciler, spec: remove complete pods

We leave the default `failedJobsHistoryLimit` - 1 - to enable the
user to debug the reconciler in case it fails.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>

* reconciler, spec: do not retry on error

Since a new pod will be scheduled in 5 minutes, it is safer to just
wait for the new reconcile cycle than insist on reconciling right
now.

This is important to prevent the cluster to clog further on
constrained clusters.

Signed-off-by: Miguel Duarte Barroso <mdbarroso@redhat.com>
